Only load routes of the app which is requested

MorrisJobke opened the pull request:
This removes ~70-90 ms on every request that involves a URLGenerator (in this case the stats are from the files default view). The total request too ~800ms so this is around 10%.

Before:
![bildschirmfoto 2019-01-21 um 12 00 44](https://user-images.githubusercontent.com/245432/51470571-91bceb80-1d74-11e9-9deb-075a37bd4e39.png)

After:
![bildschirmfoto 2019-01-21 um 12 01 24](https://user-images.githubusercontent.com/245432/51470572-92558200-1d74-11e9-85cc-b66656a08d53.png)

Diff:
![bildschirmfoto 2019-01-21 um 12 01 45](https://user-images.githubusercontent.com/245432/51470574-92558200-1d74-11e9-9d79-e407fd84e253.png)

MorrisJobke committed:
Only load routes of the app which is requested

* Add fallback to load all routes if needed
* Move partial loaded routes test to proper place

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

nickvergessen commented:
This will break activity app, because it has to generate links for various other Apps?

MorrisJobke commented:
> This will break activity app, because it has to generate links for various other Apps?

Not really. It will only load the app for which the URLs where requested. So it does not open all the `routes.php` files of all apps at once, but if you request `files.view.index` then only the ones from the files app. Beside that there is often caching involved for the routes + their parameters.

But if you request then `settings.user.index` it will also load this. But before it always loaded for each `generate()` request that was not cached all the routes that are available.

skjnldsv reviewed (approved):
Tested and works fine.
Could not find any bug on various apps and core testing

MorrisJobke commented:
This might cause problem with OCS routes: https://github.com/nextcloud/server/pull/13714/files#diff-0c5fff2a86ede7f85fc3215283d957fcR113

Let me fix that.

rullzer commented:
Yeah ocs is different unforntuantly.

Did you also test this with the caching router? As I assume this will then have less effect.

MorrisJobke commented:
> Did you also test this with the caching router? As I assume this will then have less effect.

It was with caching enabled, but somehow still caused the generation of the URL for each request. Maybe something we should look into separately.

I fixed the OCS stuff and added tests. This is now ready to go in here. Also the OCS routes are just a weird hack and maybe require some more generalization but this is a different topic.

MorrisJobke commented:
I know why this failed. The logout URL has the request parameter in the URL:

https://github.com/nextcloud/server/blob/bb86a8ca36a1aeb73a54339340b0d1ddc2f249c3/lib/private/legacy/user.php#L272-L277

And thus the caching does not hit in:

https://github.com/nextcloud/server/blob/c1e4f9f30563d5eda1718d716d631d6092cd4e28/lib/private/Route/CachingRouter.php#L55

icewind1991 commented:
maybe worth adding a fallback to loading all routes?

I don't know if there is any code out there that messes with the route name format

rullzer commented:
> I know why this failed. The logout URL has the request parameter in the URL:
> 
> [server/lib/private/legacy/user.php](https://github.com/nextcloud/server/blob/bb86a8ca36a1aeb73a54339340b0d1ddc2f249c3/lib/private/legacy/user.php#L272-L277)
> 
> Lines 272 to 277 in [bb86a8c](/nextcloud/server/commit/bb86a8ca36a1aeb73a54339340b0d1ddc2f249c3)
>  		$logoutUrl = $urlGenerator->linkToRouteAbsolute( 
>  			'core.login.logout', 
>  			[ 
>  				'requesttoken' => \OCP\Util::callRegister(), 
>  			] 
>  		); 
> 
> And thus the caching does not hit in:
> 
> [server/lib/private/Route/CachingRouter.php](https://github.com/nextcloud/server/blob/c1e4f9f30563d5eda1718d716d631d6092cd4e28/lib/private/Route/CachingRouter.php#L55)
> 
> Line 55 in [c1e4f9f](/nextcloud/server/commit/c1e4f9f30563d5eda1718d716d631d6092cd4e28)
>  		$key = $this->context->getHost() . '#' . $this->context->getBaseUrl() . $name . sha1(json_encode($parameters)) . (int)$absolute;

Right. Yes and it needs to be there becuase it is a default get.
Maybe we should hack the caching router to handle this case better then. As it is special anyways.

rullzer commented on lib/private/Route/Router.php:
```diff
@@ -328,7 +328,12 @@ public function getGenerator() {
+		list($appName, $other) = explode('.', $name, 3);
+		// OCS routes are prefixed with "ocs."
+		if ($appName === 'ocs') {
+			$appName = $other;
```
Is this enough for ocs?
As not all ocs routers are as clean (like there are parts that link to `/cloud/users` etc instead of a full route). But I guess it works fine because the internal name is still properly styled.

nickvergessen replied:
#ShouldBe 

Maybe we can really add a fall back in terms of "if no route was found, load the old way"

rullzer replied:
Ah my bad. I think this is all good. THe other way around is not so clean.

Like if we want to find out `/cloud/user` we have to load more apps. But this way it is clean. Because the routes are generated internally as `ocs.<app>.<controller>.<method>`.

rullzer reviewed (approved):
Added the fallback.

Lets get this in!

MorrisJobke commented:
Changes by @rullzer make sense 👍

danielkesselberg commented on lib/private/Route/Router.php:
```diff
@@ -328,12 +328,31 @@ public function getGenerator() {
 							 $parameters = [],
 							 $absolute = false) {
+
+		if (strpos($name,'.') !== false) {
```
You could add a space between , and '.

danielkesselberg commented on lib/private/Route/Router.php:
```diff
@@ -328,12 +328,31 @@ public function getGenerator() {
+			}
+			$this->loadRoutes($appName);
+		} else {
+			$this->loadRoutes();
```
If you load all routes here but `generate` fails (not sure if this is possible) you load all routes again (348+) and try again.

MorrisJobke replied:
True - the lines 343 to 353 should go in the block before this `else` branch.

MorrisJobke replied:
Let me do that.

kesselb reviewed (approved):
:+1:

rullzer commented:
CI is not happy but I do not get why...

https://drone.nextcloud.com/nextcloud/server/15482/110

MorrisJobke commented:
> CI is not happy but I do not get why...

Me neither - I rebased. The tests work totally fine locally :/

danielkesselberg commented:
https://github.com/nextcloud/server/blob/be47d1b383311edc705a2079deb98513f8ac6c30/lib/private/Route/Router.php#L143

Route files are not loaded because the apps are not enabled. Test fails for me on master and this branch :thinking:

MorrisJobke commented:
> Route files are not loaded because the apps are not enabled. Test fails for me on master and this branch 🤔

But the files app can't be disable 🤯

nickvergessen commented:
> But the files app can't be disable 🤯

Right, but it still needs to be loaded. Anyway tests are also fine here. Rebase to see if it was a temporary thing?

MorrisJobke commented:
> Right, but it still needs to be loaded. Anyway tests are also fine here. Rebase to see if it was a temporary thing?

I already rebased twice ... same outcome :(

MorrisJobke commented:
Rebased to trigger CI again.

skjnldsv commented:
To merge?

MorrisJobke commented:
> To merge?

CI failures 🙈

kesselb commented:
Fixed the failing tests by adding `@group DB`: https://github.com/nextcloud/server/pull/13712/commits/c9b3e7accc0318dae8bc7821e42d61e31d05c120

I have no idea why we need a database connection for loading the routes of a single app but not for loading the routes for all apps :confused:

kesselb commented:
![image](https://user-images.githubusercontent.com/3902676/65823225-ae3deb80-e252-11e9-94ad-c15ee065bcac.png)

Crap! Now another tests fails. The failing test runs without database. DI tries to initialize the files app and `$resourceManager = $container->query(IManager::class);` has `IDBConnection $connection` as dependency.

https://github.com/nextcloud/server/blob/357263a70bf56e351e0d292c6144f998dad7f079/lib/private/ServerContainer.php#L95-L113

Master: `isset($this->appContainers[$namespace])` is true.
This PR: `new $applicationClassName();` tries to initialize the files app.

Adding  `@group DB` to `OCA\Files\Tests\Activity\Setting\GenericTest` seems to be a workaround instead a fix.

rullzer commented on lib/private/Route/Router.php:
```diff
@@ -350,7 +365,7 @@ public function generate($name,
 	 */
 	private function requireRouteFile($file, $appName) {
-		$this->setupRoutes(include_once $file, $appName);
+		$this->setupRoutes(include $file, $appName);
```
Why did this change? The include once is there so that if the routing file is required two time it is just empty...

MorrisJobke replied:
Fixed

MorrisJobke replied:
The change for the routes to be loaded via include instead of include_once is to allow it to properl load them in the unit tests as they re-initiate the routes during the same execution cycle.

MorrisJobke replied:
I removed this and adjusted the tests. They pass now locally.

MorrisJobke commented:
The change for the routes to be loaded via include instead of include_once is to allow it to properl load them in the unit tests as they re-initiate the routes during the same execution cycle.

MorrisJobke commented:
Should be ready for review now :)

skjnldsv reviewed (approved):
:elephant:

rullzer commented:
> The change for the routes to be loaded via include instead of include_once is to allow it to properl load them in the unit tests as they re-initiate the routes during the same execution cycle.

So we modify the code to make the test pass? That doesn't sound right. From my understanding include once is correct here.

MorrisJobke commented:
> > The change for the routes to be loaded via include instead of include_once is to allow it to properl load them in the unit tests as they re-initiate the routes during the same execution cycle.
> 
> So we modify the code to make the test pass? That doesn't sound right. From my understanding include once is correct here.

Not in the context of changed behavior:

* request `files.SOMETHING`
* load only the `files` routes
* return route
* request `files_SOMETHING`
* can't determine the app and thus need to load all routes
* load all (this includes the `files` routes again) and because it was built with "it loads anything at once anyways" it is not meant for merging routes it will load all and overwrite the existing stuff in the "load all" step

But I'm fine with refine this and change that. On the other hand: we need to re-profile, because the causing issue seem to be solved as the timing doesn't improve that heavily anymore for me locally.

MorrisJobke commented:
> So we modify the code to make the test pass? That doesn't sound right. From my understanding include once is correct here.

I reverted the change and instead streamlined the tests a bit (it initialized once the router per test case which then hits the limitations of this implementation due to the `include_once` - now the router is initialized only once).

MorrisJobke committed:
Properly inject IRouter into URLGenerator to properly encapsulate tests

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
I now properly injected the `IRouter` into the `IURLGenerator`. Now the tests are also properly separated.

MorrisJobke commented:
CI failures are unrelated.
